mteb/evaluation/evaluators/InstructionRetrievalEvaluator.py

```diff
@@ -0,0 +1,263 @@
+            
+        logger.info("Encoding Queries...")
+        query_ids = list(queries.keys())
+        self.results = {qid: {} for qid in query_ids}
+        queries = [queries[qid] for qid in queries]
+        query_embeddings = self.model.encode_queries(
+            queries, batch_size=self.batch_size, show_progress_bar=self.show_progress_bar, convert_to_tensor=self.convert_to_tensor, instructions=instructions)
+          
+        logger.info("Sorting Corpus by document length (Longest first)...")
+
+        corpus_ids = sorted(corpus, key=lambda k: len(corpus[k].get("title", "") + corpus[k].get("text", "")), reverse=True)
+        corpus = [corpus[cid] for cid in corpus_ids]
+
+        logger.info("Encoding Corpus in batches... Warning: This might take a while!")
+        logger.info("Scoring Function: {} ({})".format(self.score_function_desc[score_function], score_function))
+
+        itr = range(0, len(corpus), self.corpus_chunk_size)
+        
+        results = {qid: [] for qid in query_ids}
+        for batch_num, corpus_start_idx in enumerate(itr):
+            logger.info("Encoding Batch {}/{}...".format(batch_num+1, len(itr)))
+            corpus_end_idx = min(corpus_start_idx + self.corpus_chunk_size, len(corpus))
+
+            # Encode chunk of corpus
+            if self.save_corpus_embeddings and "qid" in kwargs and len(self.corpus_embeddings[kwargs["qid"]]):
+                sub_corpus_embeddings = self.corpus_embeddings[qid][batch_num]
+            else:
+                sub_corpus_embeddings = self.model.encode_corpus(
+                    corpus[corpus_start_idx:corpus_end_idx],
+                    batch_size=self.batch_size,
+                    show_progress_bar=self.show_progress_bar, 
+                    convert_to_tensor = self.convert_to_tensor
+                )
+                if self.save_corpus_embeddings and "qid" in kwargs:
+                    self.corpus_embeddings[kwargs["qid"]].append(sub_corpus_embeddings)
+
+            # Compute similarites using either cosine-similarity or dot product
+            cos_scores = self.score_functions[score_function](query_embeddings, sub_corpus_embeddings)
+            cos_scores[torch.isnan(cos_scores)] = -1
+
+            # get all values to put in results
+            cos_scores = cos_scores.cpu().tolist()
+            for query_itr in range(len(query_embeddings)):
+                query_id = query_ids[query_itr]                  
+                for sub_corpus_id, score in enumerate(cos_scores[query_itr]):
+                    corpus_id = corpus_ids[corpus_start_idx+sub_corpus_id]
+                    results[query_id].append((corpus_id, score))
+
+
+        for qid in results:
+            for doc_id, score in results[qid]:
+                self.results[qid][doc_id] = score
+        
+        return self.results
+
+
+class DRESModel:
```

Comment:
Same as with the exact search model, I did this because we cache the document embeddings.  I could move this into AbstractRetrieval though?